### ISSUE TITLE
cmd/go/doc & tools/cmd/godoc: whitespace/indentation in the source leaks to the output

### ISSUE THREAD

**Issue by ahmetalpbalkan** [golang/go#18524]
### What version of Go are you using (`go version`)?

go1.7

### What did you do?

I have a func:

```
func (q Query) Join(inner Query,
	outerKeySelectorFn interface{},
	innerKeySelectorFn interface{},
	resultSelectorFn interface{}) Query {
```

I run `go doc Query.JoinT` command on this package.

### What did you expect to see?

I expected to see a canonicalized single-line representation of the signature (or however the cmd sees it would render better in the output).

So if I change the indentation in the source code to the following, I expect to see the output would remain consistent:

```
func (q Query) Join(inner Query, outerKeySelectorFn interface{}, innerKeySelectorFn interface{}, resultSelectorFn interface{}) Query {
```

### What did you see instead?

The exact indentation in the source code, leaks to the `go doc` and `godoc` output.

If I move all parameters to a single line above, `go doc Query.JoinT` or `godoc .` produce single line method-signature, just like in the source code:

```
$ go doc Query.JoinT
func (q Query) JoinT(inner Query, outerKeySelectorFn interface{}, innerKeySelectorFn interface{}, resultSelectorFn interface{}) Query
```

I think the `godoc`/`go doc` output should be free of how the indentation was made in the source code. This leads to inconsistent `go doc` output for my users, if I have inconsistent styling in my source code.

**Comment by bradfitz** [comment]
Or maybe this is intentional? One could argue that it's preserving the code's formatting because the author decided it was most readable that way. I'm not sure, though?

@robpike, @dsnet?

**Comment by robpike** [comment]
Go doc prints what go/format prints.

@griesemer

**Comment by dsnet** [comment]
The documentation for `go doc` says:
> Doc prints the documentation comments associated with the item identified by its
arguments (a package, const, func, type, var, or method) followed by a one-line
summary of each of the first-level items "under" that item (package-level
declarations for a package, methods for a type, etc.).
Note that the "one-line summary" only applies to elements under the top-level.

When you do `go doc Query.JoinT`, you are explicitly asking for `JoinT`, so it is not required that it be a single line. Especially if the function or method signature is especially long, I think it is reasonable for `go doc` to respect the original formatting in this case. 

When you do `go doc Query`, you are only asking about the `Query` type, so it is reasonable to expect all of the items under it to be a single line. This does not happen in Go 1.7, but was fixed in Go 1.8 with https://go-review.googlesource.com/#/c/25420/

**Comment by dsnet** [comment]
As a follow-up, it would be really hard to come up with an algorithm that knows how to properly re-create newlines and whitespace in such a way that is satisfactory. 

Try `go doc unicode categories` and see a giant list of values. Where should the newlines be inserted? Going by what the file specifies is probably the most reasonable thing to do.
